pkg/controllers/streaming/processor_controller_test.go

```diff
@@ -0,0 +1,558 @@
+/*
+Copyright 2019 the original author or authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package streaming
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"testing"
+
+	"github.com/go-logr/logr"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/projectriff/system/pkg/apis"
+	"github.com/projectriff/system/pkg/apis/build/v1alpha1"
+	corev1alpha1 "github.com/projectriff/system/pkg/apis/core/v1alpha1"
+	streamingv1alpha1 "github.com/projectriff/system/pkg/apis/streaming/v1alpha1"
+	kedav1alpha1 "github.com/projectriff/system/pkg/apis/thirdparty/keda/v1alpha1"
+	rtesting "github.com/projectriff/system/pkg/controllers/testing"
+	"github.com/projectriff/system/pkg/tracker"
+)
+
+const (
+	testNamespace      = "test-namespace"
+	testName           = "test-processor"
+	testProcessorImage = "test-processor-image"
+	testDefaultImage   = "test-image-from-template"
+	testFunction       = "test-function"
+	testFunctionImage  = "test-function-image"
+	testContainer      = "test-container"
+	testContainerImage = "test-container-image"
+	testErrorMessage   = "test error"
+)
+
+var testError = errors.New(testErrorMessage)
+
+func TestReconcile(t *testing.T) {
+	table := rtesting.Table{{
+		Name:        "processor does not exist",
+		Key:         rtesting.Key(testNamespace, testName),
+		ExpectTrack: []rtesting.TrackRequest{},
+	}, {
+		Name:             "getting processor fails",
+		Key:              rtesting.Key(testNamespace, testName),
+		ExpectTrack:      []rtesting.TrackRequest{},
+		GetHook:          getErrorOn(testNamespace, testName),
+		ExpectErr:        true,
+		ExpectErrMessage: testErrorMessage,
+	}, {
+		Name: "configMap does not exist",
+		Objects: []runtime.Object{
+			processor(testNamespace, testName),
```

Comment:
I'm not a fan of an abundance of helper methods. I'm 👍 on helpers to generate supporting objects in consistent ways that are repeated but don't need to change (like a function). The issue that creeps up is when that are subtile difference between objected under test that are generated by a helper, the changes tend to be introduced to the helper instead of being localized in the test.

We should continue to seek a good balance, but less is not inherently more here.

Reply:
`processor` produces a fixed value unless the test passes an optional function to modify the value for the test, in which case the changes are localised in the test. I think that meets the goals you mention. Probably need to discuss this to understand your concerns better.

Reply:
It seems the main issue is the possibility of clashes of helper methods for tests in the same package. Possible solutions:
1. Put each controller in its own package.
2. Use function values inside each test method rather than adding functions to the package scope.
3. Adopt a naming convention to avoid clashes.
4. Move helper methods to a controller-specific helper package.